platform/linux-generic/arch/aarch64/odp/api/abi/atomic_inlines.h

```diff
@@ -152,7 +152,64 @@ static inline int _odp_atomic_cas_acq_rel_u128(odp_atomic_u128_t *atom, odp_u128
 
 	return 0;
 }
-#else /* _ODP_LOCK_FREE_128BIT_ATOMICS */
+
+static inline void _odp_atomic_add_u32(odp_atomic_u32_t *atom, uint32_t val)
+{
+	__asm__ volatile("stadd   %w0, [%1]"
+			 :
+			 : "r" (val), "r" (atom)
+			 : "memory");
+}
+
+static inline void _odp_atomic_sub_u32(odp_atomic_u32_t *atom, uint32_t val)
+{
+	int32_t neg_val = -val;
+
+	__asm__ volatile("stadd   %w0, [%1]"
```

Comment:
I tested the current implementation (`__atomic_fetch_add()`) with up to GCC 11.1.0 and the produced code used always `ldadd`.

Reply:
We brought this up in our discussion with Arm compiler's team last week and there is an issue in letting the compiler generate an STADD instead of LDADD from `atomic_fetch_add()` directly when the return value is not used. Essentially, if we perform an `atomic_fetch_add` before an `atomic_thread_fence (acquire)`, then the add operation should not be allowed to pass the fence. But, because we've  let the LDADD decay to an STADD for an unused return value, it can now pass the one-way barrier. You go look at this thread for a more detailed explanation: https://gcc.gnu.org/legacy-ml/gcc-patches/2018-10/msg02042.html